### PR TITLE
Add EMF codec tests, fix compat bugs and crashers

### PR DESCRIPTION
**Pull request by hughbe**
Relies on #152

### CODE REVIEW

File: src/metafile.c [hunk 1]
```diff
@@ -1251,7 +1251,63 @@ update_emf_header (MetafileHeader *header, BYTE* data, int length)
 	return status;
 }
 
-static GpStatus 
+GpStatus
+gdip_read_emf_header_optionals (ENHMETAHEADER3 *header, void *pointer, ImageSource source)
```
Comment (hughbe) [inline src/metafile.c]: This algorithm is taken from https://msdn.microsoft.com/en-us/library/cc230635.aspx

Comment (akoeplinger) [reply src/metafile.c]: could you please put that link into the source?

**Comment by akoeplinger** [comment]
looks like the linux build fails

File: src/metafile.c [hunk 2]
```diff
@@ -1466,7 +1530,9 @@ GdipCreateMetafileFromFile (GDIPCONST WCHAR *file, GpMetafile **metafile)
 	
 	fp = fopen (file_name, "rb");
 	if (fp) {
-		status = gdip_get_metafile_from (fp, metafile, File);
+		if (gdip_get_metafile_from (fp, metafile, File) == Ok)
+			status = Ok;
```
Comment (akoeplinger) [inline src/metafile.c]: not sure why this change is required?

File: tests/testmetafile.c [hunk 3]
```diff
@@ -90,33 +90,40 @@ static void test_createMetafileFromEmf ()
     GpMetafile *wmfMetafile;
     GpMetafile *emfMetafile;
     GpMetafile *metafile;
+	HENHMETAFILE hWmfMetafile;
+	HENHMETAFILE hEmfMetafile;
 
     GdipCreateMetafileFromFile (wmfFilePath, &wmfMetafile);
     GdipCreateMetafileFromFile (emfFilePath, &emfMetafile);
+	GdipGetHemfFromMetafile (wmfMetafile, &hWmfMetafile);
+	GdipGetHemfFromMetafile (emfMetafile, &hEmfMetafile);
```
Comment (akoeplinger) [inline tests/testmetafile.c]: indentation mismatch (looks like some test files still use spaces...)

**Comment by akoeplinger** [comment]
Oh and please squash related changes together, it seems there are a few mixed up now.

File: src/metafile.c (hunk 2 above)
Comment (hughbe) [reply src/metafile.c]: `gdip_get_metafile_from` returns `OutOfMemory` for invalid metafiles, but `GdipCreateMetafileFromFile` returns `GenericError`
I added more tests to validate this - to be clearer, I also moved this to the 4th commit
